src/util/codec/table.rs

```diff
@@ -229,43 +229,116 @@ pub trait TableDecoder: DatumDecoder {
 
 impl<T: BytesDecoder> TableDecoder for T {}
 
-// `cut_row` cut encoded row into byte slices and return interested columns' byte slice.
-// Row layout: colID1, value1, colID2, value2, .....
-pub fn cut_row<'a>(mut data: &'a [u8], cols: &HashSet<i64>) -> Result<ValDict<'a>> {
-    // hack: HashMap will still allocate memeory when capacity is 0, need to use new instead.
-    if cols.is_empty() {
-        return Ok(HashMap::new());
-    }
-    let mut res = HashMap::with_capacity(cols.len());
-    if data.is_empty() || data.len() == 1 && data[0] == datum::NIL_FLAG {
-        return Ok(res);
-    }
-    while !data.is_empty() && res.len() < cols.len() {
-        let id = try!(data.decode_datum()).i64();
-        let (val, rem) = try!(datum::split_datum(data, false));
-        if cols.contains(&id) {
-            res.insert(id, val);
+pub struct RowColMeta {
+    offset: usize,
+    length: usize,
+}
+
+pub struct RowColsDict {
+    // data of current row
+    value: Vec<u8>,
+    // cols contains meta of each column in the format of:
+    // (col_id1,(offset1,len1)),(col_id2,(offset2,len2),...)
+    cols: HashMap<i64, RowColMeta>,
+}
+
+impl RowColMeta {
+    pub fn new(offset: usize, length: usize) -> RowColMeta {
+        RowColMeta {
+            offset: offset,
+            length: length,
         }
-        data = rem;
     }
-    Ok(res)
 }
 
-pub type ValDict<'a> = HashMap<i64, &'a [u8]>;
+impl RowColsDict {
+    fn new(cols_num: usize, val: Vec<u8>) -> RowColsDict {
+        RowColsDict {
+            value: val,
+            cols: HashMap::with_capacity(cols_num),
+        }
+    }
 
-// `cut_idx_key` cuts encoded index key into colIDs to bytes slices map.
-pub fn cut_idx_key<'a>(mut key: &'a [u8], col_ids: &[i64]) -> Result<(ValDict<'a>, &'a [u8])> {
-    key = &key[PREFIX_LEN + ID_LEN..];
-    if col_ids.is_empty() {
-        return Ok((HashMap::new(), key));
+    fn set_value(&mut self, val: Vec<u8>) {
+        self.value = val;
```

Comment:
ignore it. You build the cols first and then set value. 

But this confuses me because I just though here you just reset the value but forget to reset the cols.  Why not create the RowColsDict with `data: Vec<u8>, cols: &HashSet<i64>` directly?